################################################################################
# This subworkflow aims at processing the ChIP-seq data. The dataset can be
# found in ../input_table/data_mouse_multireads.csv.
#
# Nicolas Descostes
################################################################################



###############################################################################
# Imports
###############################################################################

import pandas
import os
from os.path import basename, splitext, commonprefix


###############################################################################
# Config Variables
###############################################################################

try:
  input_table = config["samples"]["summaryFile"]
except KeyError:
  print("The parameter \"summaryFile\" in section \"samples\" has not been \
    defined. The pipeline cannot start.")


###############################################################################
# CHECK EXISTENCE OF FILES
###############################################################################

if not os.path.isfile(input_table):
        raise IOError("File \"" + input_table + "\" (defined in " + 
            config["samples"]["summaryFile"] + ") not found.")



###############################################################################
# Functions
###############################################################################


def read_samplesTable(inputTable):
    data = pandas.read_csv(inputTable)
    # Verify column names
    if not {'sraname', 'samples', 'run', 'link', 'organism', 'name', 'input', 
        'experiment_title', 'cell_line', 'rep', 'study_name', 'library_strategy', 
        'library_layout', 'study_title', 
        'read_length_max'}.issubset(data.columns.values):
            raise KeyError("The samples file must contain the following \
                named columns: 'sraname', 'samples', 'run', 'link', 'organism',\
                'name', 'input', 'experiment_title', 'cell_line', 'rep',\
                'study_name', 'library_strategy', 'library_layout', \
                'study_title', 'read_length_max'")
    return data

def genome_files():
    link_chr1=config["genome"]["fastaChr1Link"]
    all_links = []
    all_chr = list(range(1,20)) + list('X') + list('Y')
    
    for chr in all_chr:
        all_links.append(link_chr1.replace("chromosome.1", "chromosome." + str(chr), 1))
    
    return all_links

def create_k_names(samples, multithreshold):
    result = []
    for samplename in samples:
        for k in multithreshold:
            result.append(samplename + "_trimmed_k" + k)
    return result

def create_k_namesnext(samples, multithreshold):
    result = []
    for samplename in samples:
        for k in multithreshold:
            result.append(samplename + "_trimmed_k" + k + "_sorted_nodups")
    return result

def bestmultibuild(dfsync, multithres):
    SAMPSYNC = dfsync["samples"].tolist()
    sampsync_trimbest = [samplename + "_trimmed_best_sorted_nodups" for samplename in SAMPSYNC]
    sampsync_trimmed_k = create_k_namesnext(SAMPSYNC, multithres)
    SAMPSYNCBESTMULTI = sampsync_trimbest + sampsync_trimmed_k
    return SAMPSYNCBESTMULTI

def syncbestmulti(samplesData, multithres):
    dfexp = samplesData[samplesData['input'] != 'none']
    dfexp_input_ref = [exp.split("-")[0] for exp in dfexp['input']]
    a = dfexp_input_ref
    b = samplesData['sraname'].tolist()
    exp_input_idx=[ b.index(x) if x in b else None for x in a ]
    dfinput = samplesData.iloc[exp_input_idx]
    EXPSYNCBESTMULTI = bestmultibuild(dfexp, multithres)
    INPUTSYNCBESTMULTI = bestmultibuild(dfinput, multithres)
    return EXPSYNCBESTMULTI,INPUTSYNCBESTMULTI



###############################################################################
# Reading input table
###############################################################################


samplesData = read_samplesTable(input_table)


###############################################################################
# Parameters checking
###############################################################################

# Checking organism definition
tableorg = samplesData['organism'].unique() 
testorg = ['Mus_musculus'] 
if not (testorg == tableorg).all():
    raise KeyError("All samples should belong to Mus musculus, check your \
        summaryFile.")

# Checking the library strategy
tablestrategy = samplesData['library_strategy'].unique()
teststrategy = ['ChIP-Seq']
if not (teststrategy == tablestrategy).all():
    raise KeyError("All samples should be ChIP-Seq experiments, check your \
        summaryFile.")

# Checking the library layout
tablelayout = samplesData['library_layout'].unique()
testlayoutsingle = ['SINGLE']
testlayoutpaired = ['PAIRED']

if (testlayoutpaired == tablelayout).all():
    print("\n\n ## The table is only composed of paired-end experiments ##\n")
elif (testlayoutsingle == tablelayout).all():
    print("\n\n ## The table is only composed of single-end experiments ##\n")
elif ((testlayoutsingle == tablelayout) | 
    (testlayoutpaired == tablelayout)).all():
    print("\n\n ## The table is composed of single-end and paired-end \
        experiments ##\n")
else:
    raise KeyError("All samples should be single or paired end, check your \
        summaryFile.")



###############################################################################
# Variables definition
###############################################################################

# Splitting the table into single or paired end experiments

index_single = samplesData['library_layout'] == 'SINGLE'
index_paired = samplesData['library_layout'] == 'PAIRED'
samplesData_single = samplesData[index_single]
samplesData_paired = samplesData[index_paired]

# Output files names

SINGLESAMPLES = samplesData_single['samples'].tolist()
PAIREDSAMPLES = samplesData_paired['samples'].tolist()

# For Retrieving links to download sra files

samples_single_df = pandas.DataFrame(samplesData_single).set_index("samples",
    drop=False)
samples_paired_df = pandas.DataFrame(samplesData_paired).set_index("samples",
    drop=False)

# Information regarding the genome
GENOMEID=config["genome"]["ID"]
GENOMELINKS=genome_files()
PREFIXFASTAGTF=config["genome"]["prefix"]
VERSION=config["genome"]["version"]

# Information for alignment with Bowtie2
MULTITHRESHOLD=["10", "50", "100", "150", "200", "250", "300", "350", "400"]

#  Define files names for BAM format grouping the best, k and all names
single_trimmed_best = [samplename + "_trimmed_best" for samplename in SINGLESAMPLES]
single_trimmed_k = create_k_names(SINGLESAMPLES, MULTITHRESHOLD)
SINGLEBESTMULTI = single_trimmed_best + single_trimmed_k

paired_trimmed_best = [samplename + "_trimmed_best" for samplename in PAIREDSAMPLES]
paired_trimmed_k = create_k_names(PAIREDSAMPLES, MULTITHRESHOLD)
PAIREDBESTMULTI = paired_trimmed_best + paired_trimmed_k

# Information for peak detection
QVALTHRES = ["0.001", "0.01", "0.02", "0.03", "0.04", "1e-04", "1e-05"]
MODELTYPE = ["no_model", "model_based"]

# Building table of corresponding exp/input
(sync_single_expdf, sync_single_inputdf) = syncbestmulti(samplesData_single, MULTITHRESHOLD)
(sync_paired_expdf, sync_paired_inputdf) = syncbestmulti(samplesData_paired, MULTITHRESHOLD)  


###############################################################################
# Rules
###############################################################################


rule all:
  input:
    expand("../results/data/fasta/{genome}/{prefix}.dna.chromosome.fa",
        genome = GENOMEID, prefix = PREFIXFASTAGTF),
    expand("../results/data/gtf/{genome}/{prefix}.{version}.chr.gtf",
        genome = GENOMEID, prefix = PREFIXFASTAGTF, version = VERSION),
    expand("../results/data/gff/{genome}/{prefix}.{version}.chr.gff3",
        genome = GENOMEID, prefix = PREFIXFASTAGTF, version = VERSION),
    expand("../results/data/bowtie2_index/{genome}/{prefix}.1.bt2",
        genome = GENOMEID, prefix = PREFIXFASTAGTF),
    expand("../results/data/bowtie2_index/{genome}/{prefix}.2.bt2",
        genome = GENOMEID, prefix = PREFIXFASTAGTF),
    expand("../results/data/bowtie2_index/{genome}/{prefix}.3.bt2",
        genome = GENOMEID, prefix = PREFIXFASTAGTF),
    expand("../results/data/bowtie2_index/{genome}/{prefix}.4.bt2",
        genome = GENOMEID, prefix = PREFIXFASTAGTF),
    expand("../results/data/bowtie2_index/{genome}/{prefix}.rev.1.bt2",
        genome = GENOMEID, prefix = PREFIXFASTAGTF),
    expand("../results/data/bowtie2_index/{genome}/{prefix}.rev.2.bt2",
        genome = GENOMEID, prefix = PREFIXFASTAGTF),
    #expand("../results/data/sra/single/{singleEndName}", 
    #    singleEndName = SINGLESAMPLES),
    #expand("../results/data/sra/paired/{pairedEndName}", 
    #    pairedEndName = PAIREDSAMPLES),
    #expand("../results/data/single/{singleEndName}.fastq",
    #    singleEndName = SINGLESAMPLES),
    #expand("data/paired/{pairedEndName}_1.fastq",
    #    pairedEndName = PAIREDSAMPLES),
    #expand("data/paired/{pairedEndName}_2.fastq",
    #    pairedEndName = PAIREDSAMPLES),
    #expand("../results/data/single/{singleEndName}.fastq.gz",
    #    singleEndName = SINGLESAMPLES),
    #expand("../results/data/paired/{pairedEndName}_1.fastq.gz",
    #    pairedEndName = PAIREDSAMPLES),
    #expand("../results/data/paired/{pairedEndName}_2.fastq.gz",
    #    pairedEndName = PAIREDSAMPLES)
    expand("../results/qc/fastqc/raw_fastq/single/{singleEndName}_fastqc.html",
        singleEndName = SINGLESAMPLES),
    expand("../results/qc/fastqc/raw_fastq/single/{singleEndName}_fastqc.zip",
        singleEndName = SINGLESAMPLES),
    expand("../results/qc/fastqc/raw_fastq/paired/{pairedEndName}_1_fastqc.html",
        pairedEndName = PAIREDSAMPLES),
    expand("../results/qc/fastqc/raw_fastq/paired/{pairedEndName}_1_fastqc.zip",
        pairedEndName = PAIREDSAMPLES),
    expand("../results/qc/fastqc/raw_fastq/paired/{pairedEndName}_2_fastqc.html",
        pairedEndName = PAIREDSAMPLES),
    expand("../results/qc/fastqc/raw_fastq/paired/{pairedEndName}_2_fastqc.zip",
        pairedEndName = PAIREDSAMPLES),
    #expand("../results/data/single/trimmed/{singleEndName}_trimmed.fq.gz",
    #    singleEndName = SINGLESAMPLES),
    #expand("../results/data/single/trimmed/{singleEndName}.fastq.gz_trimming_report.txt",
    #    singleEndName = SINGLESAMPLES),
    #expand("../results/data/paired/trimmed/{pairedEndName}_1_val_1.fq.gz",
    #    pairedEndName = PAIREDSAMPLES),
    #expand("../results/data/paired/trimmed/{pairedEndName}_1.fastq.gz_trimming_report.txt",
    #    pairedEndName = PAIREDSAMPLES),
    #expand("../results/data/paired/trimmed/{pairedEndName}_2_val_2.fq.gz",
    #    pairedEndName = PAIREDSAMPLES),
    #expand("../results/data/paired/trimmed/{pairedEndName}_2.fastq.gz_trimming_report.txt",
    #    pairedEndName = PAIREDSAMPLES),
    #expand("../results/qc/fastqc/trimmed_fastq/single/{singleEndName}_trimmed_fastqc.html",
    #    singleEndName = SINGLESAMPLES),
    expand("../results/qc/fastqc/trimmed_fastq/single/{singleEndName}_trimmed_fastqc.zip",
        singleEndName = SINGLESAMPLES),
    expand("../results/qc/fastqc/trimmed_fastq/paired/{pairedEndName}_1_val_1_fastqc.html",
        pairedEndName = PAIREDSAMPLES),
    expand("../results/qc/fastqc/trimmed_fastq/paired/{pairedEndName}_1_val_1_fastqc.zip",
        pairedEndName = PAIREDSAMPLES),
    expand("../results/qc/fastqc/trimmed_fastq/paired/{pairedEndName}_2_val_2_fastqc.html",
        pairedEndName = PAIREDSAMPLES),
    expand("../results/qc/fastqc/trimmed_fastq/paired/{pairedEndName}_2_val_2_fastqc.zip",
        pairedEndName = PAIREDSAMPLES),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singleEndName}_trimmed_best.sam",
    #    genome = GENOMEID, singleEndName = SINGLESAMPLES),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singleEndName}_trimmed_best.log",
    #    genome = GENOMEID, singleEndName = SINGLESAMPLES),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singleEndName}_trimmed_k{multi}.sam",
    #    genome = GENOMEID, singleEndName = SINGLESAMPLES, multi = MULTITHRESHOLD),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singleEndName}_trimmed_k{multi}.log",
    #    genome = GENOMEID, singleEndName = SINGLESAMPLES, multi = MULTITHRESHOLD),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singleEndName}_trimmed_all.sam",
    #    genome = GENOMEID, singleEndName = SINGLESAMPLES),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singleEndName}_trimmed_all.log",
    #    genome = GENOMEID, singleEndName = SINGLESAMPLES),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedEndName}_trimmed_best.sam",
    #    genome = GENOMEID, pairedEndName = PAIREDSAMPLES),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedEndName}_trimmed_best.log",
    #    genome = GENOMEID, pairedEndName = PAIREDSAMPLES),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedEndName}_trimmed_k{multi}.sam",
    #    genome = GENOMEID, pairedEndName = PAIREDSAMPLES, multi = MULTITHRESHOLD),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedEndName}_trimmed_k{multi}.log",
    #    genome = GENOMEID, pairedEndName = PAIREDSAMPLES, multi = MULTITHRESHOLD),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedEndName}_trimmed_all.sam",
    #    genome = GENOMEID, pairedEndName = PAIREDSAMPLES),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedEndName}_trimmed_all.log",
    #    genome = GENOMEID, pairedEndName = PAIREDSAMPLES),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singleEndName}_trimmed_best.bam",
    #    genome = GENOMEID, singleEndName = SINGLESAMPLES),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singleEndName}_trimmed_k{multi}.bam",
    #    genome = GENOMEID, singleEndName = SINGLESAMPLES, multi = MULTITHRESHOLD),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singleEndName}_trimmed_all.bam",
    #    genome = GENOMEID, singleEndName = SINGLESAMPLES),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedEndName}_trimmed_best.bam",
    #    genome = GENOMEID, pairedEndName = PAIREDSAMPLES),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedEndName}_trimmed_k{multi}.bam",
    #    genome = GENOMEID, pairedEndName = PAIREDSAMPLES, multi = MULTITHRESHOLD),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedEndName}_trimmed_all.bam",
    #    genome = GENOMEID, pairedEndName = PAIREDSAMPLES)
    #expand("../results/bam/single/bowtie2_results/{genome}/{singlebestmulti}_sorted.bam",
    #    genome = GENOMEID, singlebestmulti = SINGLEBESTMULTI),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedbestmulti}_sorted.bam",
    #    genome = GENOMEID, pairedbestmulti = PAIREDBESTMULTI),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singlebestmulti}_sorted.bam.bai",
    #    genome = GENOMEID, singlebestmulti = SINGLEBESTMULTI),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedbestmulti}_sorted.bam.bai",
    #    genome = GENOMEID, pairedbestmulti = PAIREDBESTMULTI),
    expand("../results/qc/bowtie2_saturation_percentmultireads/{genome}/single/{singleEndName}.png",
        genome = GENOMEID, singleEndName = SINGLESAMPLES),
    expand("../results/qc/bowtie2_saturation_percentmultireads/{genome}/single/{singleEndName}.txt",
        genome = GENOMEID, singleEndName = SINGLESAMPLES),
    expand("../results/qc/bowtie2_saturation_percentmultireads/{genome}/paired/{pairedEndName}.png",
        genome = GENOMEID, pairedEndName = PAIREDSAMPLES),
    expand("../results/qc/bowtie2_saturation_percentmultireads/{genome}/paired/{pairedEndName}.txt",
        genome = GENOMEID, pairedEndName = PAIREDSAMPLES),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singlebestmulti}_sorted_nodups.bam",
    #    genome = GENOMEID, singlebestmulti = SINGLEBESTMULTI),
    #expand("../results/bam/single/bowtie2_results/{genome}/{singlebestmulti}_sorted_nodups.txt",
    #    genome = GENOMEID, singlebestmulti = SINGLEBESTMULTI),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedbestmulti}_sorted_noDups.bam",
    #    genome = GENOMEID, pairedbestmulti = PAIREDBESTMULTI),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedbestmulti}_sorted_noDups.txt",
    #    genome = GENOMEID, pairedbestmulti = PAIREDBESTMULTI)
    #expand("../results/bam/single/bowtie2_results/{genome}/{singlebestmulti}_sorted_nodups.bam.bai",
    #    genome = GENOMEID, singlebestmulti = SINGLEBESTMULTI),
    #expand("../results/bam/paired/bowtie2_results/{genome}/{pairedbestmulti}_sorted_noDups.bam.bai",
    #    genome = GENOMEID, pairedbestmulti = PAIREDBESTMULTI),
    #expand("../results/qc/elongation_size_single_csaw/{genome}/{singlebestmulti}.txt",
    #    genome = GENOMEID, singlebestmulti = SINGLEBESTMULTI),
    expand("../results/bigwig/single/{genome}/{singlebestmulti}.bw",
        genome = GENOMEID, singlebestmulti = SINGLEBESTMULTI),
    expand("../results/bigwig/paired/{genome}/{pairedbestmulti}.bw",
        genome = GENOMEID, pairedbestmulti = PAIREDBESTMULTI),
    expand("../results/bigwig/single/{genome}/{singlebestmulti}_norm.bw",
        genome = GENOMEID, singlebestmulti = SINGLEBESTMULTI),
    expand("../results/bigwig/paired/{genome}/{pairedbestmulti}_norm.bw",
        genome = GENOMEID, pairedbestmulti = PAIREDBESTMULTI)


include: "rules/download.smk"
include: "rules/sra2fastq.smk"
include: "rules/trimming.smk"
include: "rules/qc.smk"
include: "rules/bowtie2.smk"
include: "rules/sam2bam.smk"
include: "rules/sort.smk"
#include: "rules/index_bowtie2.smk"
include: "rules/saturation.smk"
include: "rules/duplicates.smk"
include: "rules/index_bigwig.smk"
include: "rules/fragmentsize.smk"
include: "rules/bigwig.smk"
